petastorm/pytorch.py

```diff
@@ -0,0 +1,118 @@
+#  Copyright (c) 2017-2018 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from petastorm.reader import Reader, ShuffleOptions
+from petastorm.workers_pool.thread_pool import ThreadPool
+
+from torch.utils.data.dataloader import default_collate
+
+
+class DataLoader(object):
+    """
+    A data loader adaptor for ``torch.utils.data.DataLoader``.
+
+    This class iterates and returns items from the Reader in batches.
+
+    This loader can be used as a context manager, but it will terminate at the end of an epoch.
+    The context will invoke next_epoch() upon entry.
+
+    If not used as context manager, invoke the next_epoch() function at the start of each epoch, and
+    once more at the very end.
+    """
+
+    def __init__(self, dataset_url, schema_fields=None, batch_size=1, shuffle_options=None,
+                 collate_fn=default_collate, transform=None,
+                 predicate=None, rowgroup_selector=None,
+                 num_workers=0, reader_pool=None, num_epochs=1, sequence=None,
+                 training_partition=None, num_training_partitions=None,
+                 read_timeout_s=None, cache=None):
+        """
+        Initializes a data loader object, with a default collate and optional transform functions.
```

Comment:
A question: would it make sense to pass a reader instance as a parameter to the DatasetLoader instead of forwarding all arguments to it? Kinda same as pytorch's Loader takes an instance of a `Dataset`?

And on the same note, the `reader_pool` is not forwarded, which emphasizes the point :)

Reply:
Agreed! @rgruener also commented on this exact same thing. Great minds! :-)

Posting an updated commit real soon that addresses all these review comments.  Thank you so much!